src/adcm_client/base.py

```diff
@@ -443,3 +458,103 @@ def __init__(self, api: ADCMApiWrapper, path=None, path_args=None, paging=None,
                 )
             )
         super().__init__(data)
+
+
+##################################################
+#              R I C H L Y   T Y P E D
+##################################################
+
+
+def _simplify_filter(value):
+    """
+    Try to convert complex value like Enum/date/datetime to an int/str,
+    so the API will take it as a filter argument
+    """
+    if isinstance(value, (str, int)):
+        return value
+    if isinstance(value, Enum):
+        return value.value
+    if isinstance(value, datetime.date):
+        return value.strftime("%Y-%m-%d")
+    if isinstance(value, datetime.datetime):
+        return value.isoformat()
+    return value
+
+
+class RichlyTypedAPIObject(BaseAPIObject):
+    """
+    Use this class instead of `BaseAPIObject` for your objects
+    when you want to convert some fields to a complex objects
+    like Enum, datetime, etc.
+    """
+
+    def __init__(self, *args, **kwargs):
+        # for host cluster_id can be a part of a path of a filter,
+        # but since it's an int, there shouldn't be a problem anyway
+        kwargs.update({k: _simplify_filter(v) for k, v in kwargs.items() if k in self.FILTERS})
+        super().__init__(*args, **kwargs)
+
+    def _register_attrs(self):
+        super()._register_attrs()
+        try:
+            self._convert()
+        except Exception as e:  # pylint: disable=broad-except
+            warnings.warn(f'Some of the fields left was not converted due to error: {e}\n')
+
+    @abstractmethod
+    def _convert(self):
+        """Explicitly convert each field that has complex type after the data was fetched"""
+        raise NotImplementedError
+
+    def _convert_enum(self, field: str, enum_cls: Type[Enum]):
+        raw_value = getattr(self, field)
+        try:
+            setattr(self, field, enum_cls(raw_value))
+        except ValueError:
+            warnings.warn(
+                f'Failed to convert field {field} to enum {enum_cls}.\n'
+                'You might be using client version not fully compatible with ADCM version.'
+            )
+
+    def _convert_datetime(self, field: str) -> None:
+        """Convert string to datetime in iso format"""
+        raw_value = getattr(self, field)
+        # most likely DRF format
+        try:
+            setattr(
+                self,
+                field,
+                datetime.datetime.strptime(raw_value, "%Y-%m-%dT%H:%M:%S.%fZ").replace(
+                    tzinfo=datetime.timezone.utc
+                ),
```

Comment:
In my understanding, the Z on the end means this date is UTC, so we can just set "correct" timezone